### PR TITLE
Revamp tabs & follow ARIA 1.1 practices 

### LINKED ISSUE #28918: make dynamic tabs actually ARIA 1.1 practices compliant
A breaking change, but would like to squeeze this into v5 if we can: turn dynamic tabs into actual "Tabs with automatic activation" (see https://www.w3.org/TR/wai-aria-practices-1.1/examples/tabs/tabs-1/tabs.html and https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel)

- only the currently selected/active tab receives focus, all other tabs are not focusable (`tabindex="-1"` added dynamically to them)
- cursor keys cycle between tabs, setting both the `tabindex="0"` and `aria-selected="true"` to the newly active tab (and `tabindex="-1"` / `aria-selected="false"` on the now not selected one)
- mouse behavior can remain as now (though need to flip `tabindex` accordingly)
- document this in the docs, referencing specifically the "automatic activation" route (so authors aren't confused/don't wonder why using `SPACE` or whatever does nothing)

### LINKED ISSUE #31345: [Accessibility] nav tabs : tabpanel focus issues
When you select a tab the current `tabpanel` is not focusable (bootstrap v5).
With some screen readers (like NVDA, ORCA) you can't read the `tabpanel` content selected like WAI ARIA practices.
See https://www.w3.org/TR/wai-aria-practices/examples/tabs/tabs-2/tabs.html : 
> When the `tab` list contains the focus, moves focus to the next element in the `tab` sequence, which is the `tabpanel` element.

### LINKED ISSUE #34814: Tabs Component Keyboard Navigation
Hello.  Our Accessibility auditor has brought up the issue that the tab behavior in bootstrap doesn't meet their expectations.

The expected behavior:
Tab (key) - once to navigate into the group of tabs and once to navigate out of the group of tabs
↑/↓ or ←/→ - choose and activate previous/next tab.

What is happening:
Tab (key) moves into the group, and tab moves between the tabs, and the arrow keys don't function at all.

https://webaim.org/techniques/keyboard/

### PR DESCRIPTION
**Pull request by GeoSot**
It may be a bit late to, apply it.  It is almost **re-written** to focus on parent element in order to help in siblings filtering

P.S; The first commit was initialized on parent, which I found less demanding (and I prefer it), but I don't know if it is time for such major change

Closes #28918, closes #31345, closes #34814

Preview: https://deploy-preview-33079--twbs-bootstrap.netlify.app/docs/5.1/components/navs-tabs/#javascript-behavior

TODO:

- [x] `Tab` static

### CODE REVIEW

**Comment by XhmikosR** [comment]
Thanks for the PR! Personally, I don't think we should land any breaking changes anymore.

**Comment by GeoSot** [comment]
I can only agree on this. I 've cross check the tests, and it is on bootstrap team decision

**Comment by patrickhlauke** [comment]
Just getting around to checking this https://deploy-preview-33079--twbs-bootstrap.netlify.app/docs/5.0/components/navs-tabs/#javascript-behavior

This may need some additional initialisation, as currently the "make only the active tab focusable / allow cursor keys to change between them" only works after the first time a user changes the tabs

**Comment by patrickhlauke** [comment]
heads-up: ideally, not initialising on focus, but on page load itself. that was the part that stumped me (with our current JS approach). either that or we say authors need to explicitly initialise them when they're using them (same as with tooltips)

**Comment by GeoSot** [comment]
We have an opened issue for explicit initialization of tooltips/popover. So I 've added an auto-initialization on window load, to experiment with it, and if we proceed, we may discuss it a bit  more

**Comment by patrickhlauke** [comment]
going to https://deploy-preview-33079--twbs-bootstrap.netlify.app/docs/5.0/components/navs-tabs/#javascript-behavior still has all tabs by default focusable (if you start off `Tab`-bing through the page). After initialisation, only the currently active tab should be focusable.

If you set focus to one of the tabs, and *then* use cursor keys, the other tabs you land on also become non-focusable by default. but the ones you don't get to still can receive focus.

only after you go to a tabbed interface and cursor through all the tabs does it end up behaving as expected (a single tab, the active one, receives focus in the tablist)

current focus behavior

**Comment by GeoSot** [comment]
So,
*  initialize only active tabs on window load
* only active tabs are focus-able
* on keyboard right/left cycle them
* many aria attrs addons (line 225 to 264)

**Comment by patrickhlauke** [comment]
will check in a bit, but yeah basically, this is a good exemplar of how tabs should behave https://www.w3.org/TR/wai-aria-practices/examples/tabs/tabs-1/tabs.html

**Comment by patrickhlauke** [comment]
The intialisation now looks good, and I see you've also made the content panels themselves focusable now (which addresses https://github.com/twbs/bootstrap/issues/31345)

Worth checking if it also handles disabled tabs correctly (i.e. if a tab is disabled, it should not be focusable). And thinking how to handle edge cases like a tab panel where no tab is initially visible (which is not really a sensible thing, but the script should be able to handle it somehow, as otherwise there's a risk none of the tabs would then be focusable and the whole thing would become unusable).

**Comment by GeoSot** [comment]
The edge case that we have a tab `active & disabled`, I think is out of our scope. 
For now if a tab is invisible, gets the aria attributes, so when it become visible, will have the proper functionality.
The case that may need to be fix is t the keyboard navigation, when we have disabled elements.

**Comment by patrickhlauke** [comment]
I lost sight of this... what's the state of play here @GeoSot ? are there outstanding parts / things that still need to be explored?

**Comment by GeoSot** [comment]
@patrickhlauke we are almost on the same spot.
>The edge case that we have a tab active & disabled, I think is out of our scope.
For now if a tab is invisible, gets the aria attributes, so when it become visible, will have the proper functionality.
The case that may need to be fix is t the keyboard navigation, when we have disabled elements.

The only difference are some changes after a help session with @alpadev , where I've changed some selectors and the way the component searches for nested selectors, to cover the co-existent  dropdown with an inner tab element. 

PS: IMO dropdown messes a bit the mark-up,. You can see and example [on our tests](https://github.com/twbs/bootstrap/blob/bf0936748602c8109fd916c64b4560799fa1c3f8/js/tests/unit/tab.spec.js#L460), where the [tab trigger](https://github.com/twbs/bootstrap/blob/bf0936748602c8109fd916c64b4560799fa1c3f8/js/tests/unit/tab.spec.js#L468) is two levels deeper than the other tab-triggers

**Comment by patrickhlauke** [comment]
> PS: IMO dropdown messes a bit the mark-up,. You can see and example [on our tests](https://github.com/twbs/bootstrap/blob/bf0936748602c8109fd916c64b4560799fa1c3f8/js/tests/unit/tab.spec.js#L460), where the [tab trigger](https://github.com/twbs/bootstrap/blob/bf0936748602c8109fd916c64b4560799fa1c3f8/js/tests/unit/tab.spec.js#L468) is two levels deeper than the other tab-triggers

tabs with dropdowns were always an abomination. we already strongly suggest that authors SHOULD NOT use them and we removed any examples to that effect https://getbootstrap.com/docs/5.0/components/navs-tabs/#javascript-behavior ... i guess it's too late to completely gut out the code that supports them (as that code simply can't be made to work with proper ARIA tab model that we're implementing here) @XhmikosR @mdo

**Comment by XhmikosR** [comment]
So, are we sure the new behavior is not a breaking change? I compared https://twbs-bootstrap.netlify.app/docs/5.1/components/navs-tabs/#javascript-behavior and https://deploy-preview-33079--twbs-bootstrap.netlify.app/docs/5.1/components/navs-tabs/#javascript-behavior with the keyboard tab and what's tabable/selectable is different with these changes.

**Comment by patrickhlauke** [comment]
@XhmikosR yes it is a breaking change in behaviour, because our previous/current behaviour doesn't quite match the expected behaviour that we should have had all along (i.e. if you expose the tab interface as an ARIA tabpanel, you should ideally follow the actual expected interaction method for that kind of interface, as in https://www.w3.org/TR/wai-aria-practices/examples/tabs/tabs-1/tabs.html)

so it's breaking in the sense of "it corrects a longstanding incorrect behaviour"

the only alternative i can see to making it "non-breaking" would be to add the correct behaviour as an opt-in, and keept the broken behvaviour as the default. or defer this to when we can next do a breaking change...

**Comment by patrickhlauke** [comment]
sorry, my fat finger pressed the close button by accident...

**Comment by GeoSot** [comment]
Made a 'minor' change to support existing functionality, till next major version

**Comment by XhmikosR** [comment]
@patrickhlauke I thought about it and I say that we could try landing this even though it's a change in behavior. I mean, if someone is using the keyboard, my guess is that they'd find the new behavior normal, while the previous one wasn't.

Let us know what you think and if you have any suggestions about the changes in the docs (No1 etc) feel free to push them.

File: js/src/tab.js [hunk 1]
```diff
@@ -17,158 +17,262 @@ import BaseComponent from './base-component'
...
 
-      element.setAttribute('aria-expanded', true)
+    if (!target) {
+      return
     }
 
-    if (callback) {
-      callback()
+    this._setAttributeIfNotExists(target, 'tabindex', '0')
+    this._setAttributeIfNotExists(target, 'role', 'tabpanel')
```
Comment (patrickhlauke) [inline js/src/tab.js:216]: I would actually leave this aspect out, leaving it up to the users to set their tabpanels to be focusable in their own markup. reason being that this *can* be quite surprising/counterproductive, particularly if the tabpanel itself already has a focusable element right at the start - see the `Tab` behaviour https://www.w3.org/TR/wai-aria-practices-1.2/#keyboard-interaction-21 second bullet

> When the tab list contains the focus, moves focus to the next element in the page tab sequence outside the tablist, which is the tabpanel *unless the first element containing meaningful content inside the tabpanel is focusable.*

We can add something in the documentation text itself strongly advising authors to make them focusable unless they have a reason not to - that gives them some options

File: js/src/tab.js [hunk 2]
```diff
@@ -17,158 +17,262 @@ import BaseComponent from './base-component'
...
-    element.classList.add(CLASS_NAME_ACTIVE)
-    if (element.getAttribute('role') === 'tab') {
-      element.setAttribute('aria-selected', true)
+    this._queueCallback(complete, element, isAnimated)
+  }
+
+  _keydown(event) {
+    if (!([ARROW_LEFT_KEY, ARROW_RIGHT_KEY].includes(event.key))) {
+      return
+    }
```
Comment (patrickhlauke) [inline js/src/tab.js:164]: See note 2 at the end of https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-21

> When a tab list has its [aria-orientation](https://www.w3.org/TR/wai-aria-1.2/#aria-orientation) set to vertical:
> * Down Arrow performs as Right Arrow is described above.
> * Up Arrow performs as Left Arrow is described above.

For simplicity's sake, and because assistive tech don't all consistently announce `aria-orientation` anyway (as far as the user is concerned, they're in a tab list, with no hint whether it's horizontal or vertical), I'd actually say it's pointless checking explicitly what the orientation is, and instead listening to and reacting the same to *both* down/right and up/left the same way (i.e. that you can go to previous with up/left and next with down/right regardless of the orientation)

**Review by patrickhlauke — CHANGES_REQUESTED** [review]

**Comment by patrickhlauke** [comment]
sorry, some last minute picky things. if they're ok to implement and make sense, i can think of a bit of extra prose for the documentation (including the advice for authors to make tab panels focusable unless they have a good reason not to, and updating the link to point to ARIA Practices 1.2 which is the more up-to-date one)

File: js/src/tab.js (hunk 2 above)
Comment (GeoSot) [reply js/src/tab.js:164]: DONE ;)

**Comment by GeoSot** [comment]
@patrickhlauke  I am not sure I have understand the
> I would actually leave this aspect out, leaving it up to the users to set their tabpanels to be focusable in their own markup. reason being that this can be quite surprising/counterproductive, particularly if the tabpanel itself already has a focusable element right at the start - see the Tab behaviour https://www.w3.org/TR/wai-aria-practices-1.2/#keyboard-interaction-21 second bullet

You suggest to handle panel aria attributes , only if they already exist, right?

**Comment by patrickhlauke** [comment]
> @patrickhlauke I am not sure I have understand the
> 
> > I would actually leave this aspect out, leaving it up to the users to set their tabpanels to be focusable in their own markup. reason being that this can be quite surprising/counterproductive, particularly if the tabpanel itself already has a focusable element right at the start - see the Tab behaviour https://www.w3.org/TR/wai-aria-practices-1.2/#keyboard-interaction-21 second bullet
> 
> You suggest to handle panel aria attributes , only if they already exist, right?

no, i mean just don't set `tabindex` at all on it in the initialisation. if authors do want that, they can add `tabindex="0"` to their markup themselves (and we'll mention it in the docs text). it's currently a bit of a controversial/awkward part of the pattern in the ARIA Practices.

**Comment by patrickhlauke** [comment]
the cursor up/down works nicely now. however, one minor thing: i think you need to `preventDefault()`, as otherwise it leads to scrolling up/down.

using up/down cursor on tab list changes active tab AND scrolls the page

File: js/src/tab.js (hunk 1 above)
Comment (patrickhlauke) [reply js/src/tab.js:218]: just the `tabindex` one (sorry, i see i missed the correct line by one when adding my comment). leave the `role` and `aria-labelledby` ones there

**Comment by patrickhlauke** [comment]
added some more documentation text. perhaps it's a bit wordy/too detailed? not sure...

**Review by mdo — APPROVED** [review]
Looks good to me, but would love @patrickhlauke's eyes one more time.

**Comment by patrickhlauke** [comment]
> Looks good to me, but would love @patrickhlauke's eyes one more time.

works as it should (mirrors structure/behaviour of https://www.w3.org/TR/wai-aria-practices-1.2/examples/tabs/tabs-1/tabs.html). if @mdo is happy with the (slightly wordy) additions to the documentation I made, this should be good to go.

**Review by patrickhlauke — APPROVED** [review]

**Comment by patrickhlauke** [comment]
@mdo your latest change to the docs is confusing. the accessibility information only relates to the javascript behaviour/plugin. so it should be a subsection of THAT, rather than coming before it

**Review by patrickhlauke — CHANGES_REQUESTED** [review]
The new "Accessibility" bit is only relevant to the plugin, so should come as a subsection of that

**Review by patrickhlauke — APPROVED** [review]

**Comment by patrickhlauke** [comment]
Lovely, thanks @mdo

**Comment by GeoSot** [comment]
Thank you too guys ❤️

**Comment by patrickhlauke** [comment]
Closing the loop, I filed some observations back to ARIA Practices https://github.com/w3c/aria-practices/issues/2281
